libraries/chain/controller.cpp

```diff
@@ -4316,8 +4316,77 @@ struct controller_impl {
       }
    }
 
+   // This is only used during Savanna transition, which is a one-time occurrance
+   // and the number of Transition block is small.
```

Comment:
```suggestion
   // This is only used during Savanna transition, which is a one-time occurrence
   // and the number of Transition blocks is small.
```

Reply:
yep, depends on your definition of "small"

Reply:
I'm a little worried about doing 200+ `calculate_merkle` in in a row... couldn't that take more than the block time?

Reply:
Good point. We could cache `action_mroot_savanna` instead of `action_receipt_digests_savanna` in legacy block state. That would add some cost when building legacy block state, but the cost is amortized.

Reply:
Yes, I think that's clearly what we should do.

Reply:
Will do it shortly.